File: Java/BubbleSort.java
import java.util.Arrays;

/**
 * BubbleSort - Implementation of the Bubble Sort Algorithm
 * 
 * <p>Bubble Sort is a simple sorting algorithm that repeatedly steps through
 * the list, compares adjacent elements, and swaps them if they are in the
 * wrong order. The pass through the list is repeated until the list is sorted.</p>
 * 
 * <h2>Algorithm Characteristics:</h2>
 * <ul>
 *   <li><b>Time Complexity:</b> O(n²) worst/average case, O(n) best case</li>
 *   <li><b>Space Complexity:</b> O(1) - sorts in-place</li>
 *   <li><b>Stability:</b> Stable - maintains relative order of equal elements</li>
 *   <li><b>Method:</b> Comparison-based sorting</li>
 * </ul>
 * 
 * <h2>How It Works:</h2>
 * <p>The algorithm makes multiple passes through the array. In each pass:</p>
 * <ol>
 *   <li>Compare adjacent pairs of elements</li>
 *   <li>Swap them if they are out of order</li>
 *   <li>After each pass, the largest unsorted element "bubbles up" to its correct position</li>
 *   <li>Continue until no swaps are needed (array is sorted)</li>
 * </ol>
 * 
 * <h2>Optimization:</h2>
 * <p>This implementation includes an optimization that detects when the array
 * is already sorted and terminates early, achieving O(n) time in the best case.</p>
 * 
 * @author Code-Contribution Community
 * @version 1.1
 * @since 2025-10-15
 */
public class BubbleSort {

    /**
     * Sorts an array of integers in ascending order using the Bubble Sort algorithm.
     * 
     * <p>This method modifies the input array in-place. It uses an optimized
     * version of bubble sort that tracks whether any swaps were made in each pass.
     * If no swaps occur, the array is already sorted and the algorithm terminates early.</p>
     * 
     * @param arr The array of integers to be sorted (modified in-place)
     * 
     * @throws NullPointerException if the array is null
     * 
     * @example
     * <pre>{@code
     * int[] numbers = {5, 2, 8, 1, 9};
     * BubbleSort.bubbleSort(numbers);
     * // numbers is now {1, 2, 5, 8, 9}
     * }</pre>
     * 
     * @see #main(String[])
     */
    public static void bubbleSort(int[] arr) {
        boolean swapped;
        
        // Outer loop: make n-1 passes through the array
        for (int i = 0; i < arr.length - 1; i++) {
            swapped = false;
            
            // Inner loop: compare and swap adjacent elements
            // After i passes, the last i elements are already in place
            for (int j = 0; j < arr.length - i - 1; j++) {
                // Compare adjacent elements
                if (arr[j] > arr[j + 1]) {
                    // Swap if they're in wrong order
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    swapped = true;
                }
            }
            
            // Optimization: if no swaps occurred, array is sorted
            if (!swapped) break;
        }
    }

    /**
     * Main method demonstrating the usage of Bubble Sort.
     * 
     * <p>Creates a sample array, sorts it using bubble sort,
     * and prints the sorted result to the console.</p>
     * 
     * @param args Command-line arguments (not used)
     * 
     * @example
     * Output for the given array:
     * <pre>
     * 0 1 1 3 4 5 6 8 10 12
     * </pre>
     */
    public static void main(String[] args) {
        // Sample array to demonstrate sorting
        int[] a = {1, 5, 6, 8, 10, 4, 12, 3, 0, 1};
        
        System.out.println("Original array: " + Arrays.toString(a));
        
        // Sort the array
        bubbleSort(a);
        
        System.out.println("Sorted array:   " + Arrays.toString(a));
        
        // Alternative: print without Arrays.toString()
        System.out.print("Elements: ");
        for (int i = 0; i < a.length; i++) {
            System.out.print(a[i] + " ");
        }
        System.out.println();
    }
}

File: C++/bubbleSort.cpp
/**
 * @file bubbleSort.cpp
 * @brief Implementation of the Bubble Sort algorithm in C++
 * 
 * This program demonstrates the bubble sort algorithm, one of the simplest
 * sorting algorithms. It works by repeatedly swapping adjacent elements if
 * they are in the wrong order, causing larger elements to "bubble" to the end.
 * 
 * Algorithm Explanation:
 * Bubble sort compares each pair of adjacent items and swaps them if they are
 * in the wrong order. The pass through the list is repeated until no swaps are
 * needed, which indicates that the list is sorted.
 * 
 * Complexity Analysis:
 * - Time Complexity (Best):    O(n)   - Already sorted array
 * - Time Complexity (Average): O(n²)  - Random order
 * - Time Complexity (Worst):   O(n²)  - Reverse sorted array
 * - Space Complexity:          O(1)   - In-place sorting
 * 
 * Properties:
 * - Stable: Yes (maintains relative order of equal elements)
 * - In-place: Yes (requires only constant O(1) extra space)
 * - Adaptive: Can be optimized to stop early if array becomes sorted
 * 
 * Compilation: g++ bubbleSort.cpp -o bubbleSort
 * Execution:   ./bubbleSort
 * 
 * @author Code-Contribution Community
 * @version 1.0
 * @date October 2025
 */

#include <iostream>
using namespace std;

/**
 * @brief Sorts an array using the bubble sort algorithm
 * 
 * This function implements the bubble sort algorithm to sort an integer array
 * in ascending order. It uses nested loops to compare adjacent elements and
 * swap them if they are in the wrong order.
 * 
 * The outer loop runs (size-1) times, and in each iteration, the largest
 * unsorted element "bubbles up" to its correct position at the end.
 * 
 * @param arr The integer array to be sorted
 * @param size The number of elements in the array
 * 
 * @note This function modifies the original array in-place
 * @note The algorithm is stable and maintains the relative order of equal elements
 * 
 * @complexity O(n²) time complexity, O(1) space complexity
 * 
 * Example:
 * @code
 * int arr[] = {64, 34, 25, 12, 22};
 * bubbleSort(arr, 5);
 * // arr is now: {12, 22, 25, 34, 64}
 * @endcode
 */
void bubbleSort(int arr[], int size) {
    // Outer loop: represents the number of passes
    // After each pass, one more element is in its correct position
    for (int step = 0; step < size - 1; ++step) {
        
        // Flag to optimize: if no swaps occur in a pass, array is sorted
        bool swapped = false;
        
        // Inner loop: compares adjacent elements
        // The range decreases with each pass since last elements are already sorted
        for (int i = 0; i < size - step - 1; ++i) {
            
            // Compare adjacent elements
            if (arr[i] > arr[i + 1]) {
                // Swap if elements are in wrong order
                int temp = arr[i];
                arr[i] = arr[i + 1];
                arr[i + 1] = temp;
                swapped = true;
            }
        }
        
        // Optimization: if no swaps occurred, array is already sorted
        if (!swapped) {
            break;
        }
    }
}

/**
 * @brief Prints the elements of an integer array
 * 
 * This utility function displays all elements of an array in a formatted manner,
 * with each element separated by a space.
 * 
 * @param arr The integer array to be printed
 * @param size The number of elements in the array
 * 
 * Output format: " element1 element2 element3 ..."
 */
void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++) {
        cout << " " << arr[i];
    }
    cout << endl;
}

/**
 * @brief Main function demonstrating bubble sort
 * 
 * Creates a sample array, sorts it using bubble sort, and displays
 * both the original and sorted arrays.
 * 
 * @return 0 on successful execution
 */
int main() {
    // Sample array to demonstrate bubble sort
    int arr[] = {64, 34, 25, 12, 22, 11, 90};
    
    // Calculate the number of elements in the array
    int N = sizeof(arr) / sizeof(arr[0]);
    
    // Display original array
    cout << "Original array:\n";
    printArray(arr, N);
    
    // Sort the array using bubble sort
    bubbleSort(arr, N);
    
    // Display sorted array
    cout << "\nSorted array:\n";
    printArray(arr, N);
    
    return 0;
}